### PR TITLE
Add party to participant mappings as submission inputs

### PR DESCRIPTION
**Pull request by dajmaki**
This adds party to participant mappings as inputs to kvutils submissions.
Needed when doing a privacy-aware transformation of the resulting log entry.

Also added InsertOrdSet and made sure all DamlSubmission inputs are in
deterministic order.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+final case object EmptyInsertOrdSet extends InsertOrdSet[Any] {
+  override def _items = Queue[Any]()
+  override def _hashSet = HashSet[Any]()
+}
+
+final case class NonEmptyInsertOrdSet[T](
+    override val _items: Queue[T],
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]: You can create an empty  `NonEmtpyInsertOrderSet`. I found that pretty confussing.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  *
+  * Implemented as (Queue[T], HashSet[T]).
+  * Asymptotics:
+  *  get: O(1)
+  *  insert: O(1)
+  *  remove: O(n)
+  */
+import scala.collection.immutable.{HashSet, Set, Queue}
+
+sealed abstract class InsertOrdSet[T] extends Set[T] {
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
final class InsertOrdSet[T](_items: Queue[T], _hashSet: HashSet[T]) extends Set[T] {
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+final case class NonEmptyInsertOrdSet[T](
+    override val _items: Queue[T],
+    override val _hashSet: HashSet[T])
+    extends InsertOrdSet[T]
+
+object InsertOrdSet {
+  def empty[T] = EmptyInsertOrdSet.asInstanceOf[InsertOrdSet[T]]
+
+  def fromSeq[T](s: Seq[T]): InsertOrdSet[T] =
+    NonEmptyInsertOrdSet(Queue(s.reverse: _*), HashSet(s: _*))
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
    new InsertOrdSet(Queue(s.reverse: _*), HashSet(s: _*))
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala (hunk 2 above)
Comment (S11001001) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]: And moreover

```suggestion
sealed abstract class InsertOrdSet[T] extends AbstractSet[T] with Set[T] with GenericSetTemplate[T, InsertOrdSet] with SetLike[T, InsertOrdSet[T]] with Serializable {
```

Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]: You should problably not use case class for anything like that where you have to maintain invariant between the  This _items and _hahsSet can be broken by using the `copy` method. Furhtermore those internal fields should problably not leak outise. 

I think it is betwer to follow the approach of ImmArray, haveing a final class with private constructor. 
I added some suggestion.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  override def _hashSet = HashSet[Any]()
+}
+
+final case class NonEmptyInsertOrdSet[T](
+    override val _items: Queue[T],
+    override val _hashSet: HashSet[T])
+    extends InsertOrdSet[T]
+
+object InsertOrdSet {
+  def empty[T] = EmptyInsertOrdSet.asInstanceOf[InsertOrdSet[T]]
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
  private val Empty =  new InsertOrdSet(Queue.empty, HashSet.empty)
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  override def +(elem: T): InsertOrdSet[T] =
+    if (_hashSet.contains(elem))
+      this
+    else
+      NonEmptyInsertOrdSet(
+        elem +: _items,
+        _hashSet + elem
+      )
+
+  override def -(elem: T): InsertOrdSet[T] =
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    NonEmptyInsertOrdSet(
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      _items.filter(elem2 => elem != elem2),
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      _hashSet - elem
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    )
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+}
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+final case object EmptyInsertOrdSet extends InsertOrdSet[Any] {
+  override def _items = Queue[Any]()
+  override def _hashSet = HashSet[Any]()
+}
+
+final case class NonEmptyInsertOrdSet[T](
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    override val _items: Queue[T],
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    override val _hashSet: HashSet[T])
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    extends InsertOrdSet[T]
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+      )
+
+  override def -(elem: T): InsertOrdSet[T] =
+    NonEmptyInsertOrdSet(
+      _items.filter(elem2 => elem != elem2),
+      _hashSet - elem
+    )
+}
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion

```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+final case object EmptyInsertOrdSet extends InsertOrdSet[Any] {
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+  override def _items = Queue[Any]()
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+  override def _hashSet = HashSet[Any]()
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+}
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala (hunk 6 above)
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
    new InsertOrdSet(
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+    _items.reverseIterator
+
+  override def contains(elem: T): Boolean =
+    _hashSet.contains(elem)
+
+  override def +(elem: T): InsertOrdSet[T] =
+    if (_hashSet.contains(elem))
+      this
+    else
+      NonEmptyInsertOrdSet(
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdSet.scala]:
```suggestion
      new InsertOrdSet(
```

**Comment by dajmaki** [comment]
already cleaned it up :)

**Pull request merged**